pkg/download/download.go

```diff
@@ -0,0 +1,168 @@
+package download
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/dynatrace-oss/dynatrace-monitoring-as-code/pkg/api"
+	"github.com/dynatrace-oss/dynatrace-monitoring-as-code/pkg/download/jsoncreator"
+	"github.com/dynatrace-oss/dynatrace-monitoring-as-code/pkg/download/yamlcreator"
+	"github.com/dynatrace-oss/dynatrace-monitoring-as-code/pkg/environment"
+	"github.com/dynatrace-oss/dynatrace-monitoring-as-code/pkg/rest"
+	"github.com/dynatrace-oss/dynatrace-monitoring-as-code/pkg/util"
+	"github.com/dynatrace-oss/dynatrace-monitoring-as-code/pkg/util/files"
+)
+
+var baseProjectName = "" //for folder naming
+var cont = 0
+
+//GetConfigs Entry point that retrieves the specified configurations from a Dynatrace tenant
+func GetConfigs(environments map[string]environment.Environment, path string, downloadSpecificAPI string) (statusCode int) {
+
+	list, err := getAPIList(downloadSpecificAPI)
+	if err != nil {
+		util.Log.Fatal("The API list contains invalid values. Run monaco command to see the available options\n" + err.Error())
+		statusCode = -1
+		return statusCode
+	}
+	downloadErrors := make(map[string]error)
+
+	for _, environment := range environments {
+
+		//download configs for each environment
+		err = downloadConfigFromEnvironment(environment, path, list)
+		if err != nil {
+			downloadErrors[environment.GetId()] = err
+		}
+	}
+	statusCode = publishErrors(downloadErrors)
+	return statusCode
+}
+
+func publishErrors(errors map[string]error) (statusCode int) {
+	for environment, err := range errors {
+		util.Log.Error("Download to %s failed with error %s\n", environment, err)
+		statusCode = -1
+	}
+	if statusCode == -1 {
+		return statusCode
+	}
+	return 0
+}
+
+//returns the list of API filter if the download specific flag is used, otherwise returns all the API's
+func getAPIList(downloadSpecificAPI string) (map[string]api.Api, error) {
```

Comment:
It would make sense to return an array of errors here:

```
func getAPIList(downloadSpecificAPI string) (map[string]api.Api, []error) {
```

The concatenation of error messages (line 76) is probably not the most user-friendly approach for logging the errors raised in this method. It would lead to error messages like the following:

```
2020-12-03 11:12:09 ERROR Value foo is not a valid API name.
 Value bar is not a valid API name.
```

More readable would be:

```
2020-12-03 11:12:09 ERROR Value foo is not a valid API name.
2020-12-03 11:12:09 ERROR Value bar is not a valid API name.
```

Reply:
sure, looks a lot better, implemented []error array instead of the string concat